lms/djangoapps/course_api/api.py

```diff
@@ -266,3 +267,79 @@ def get_course_run_url(request, course_id):
+
+
+def get_course_member_queryset(course_key, include_students=True, access_roles=None, prefetch_user_course_roles=False):
+    """
+    Returns a User queryset that filters all users related to a course. For example - Students, Teachers, Staffs etc.
+
+    Examples:
+        - Get all members:
+            get_course_member_queryset(course_key)
+        - Get only students excluding staffs:
+            get_course_member_queryset(course_key, include_students=True, access_roles=[])
+        - Get only staffs excluding students:
+            get_course_member_queryset(course_key, include_students=False, access_roles=None)
+        - Get only instructors:
+            get_course_member_queryset(course_key, include_students=False, access_roles=['instructor'])
+        - Get instructors & students:
+            get_course_member_queryset(course_key, include_students=True, access_roles=['instructor'])
+        - Get all members and prefetch course related roles:
+            get_course_member_queryset(course_key, prefetch_user_course_roles=True)
+
+    Arguments:
+        course_key (CourseKey): the CourseKey for the course
+        include_students: Wether or not to include students,
+        access_roles:
+            accepts an array of string course access roles.
+            If None provided, it includes all roles.
+        prefetch_user_course_roles:
+            prefetches CourseAccessRole & CourseEnrollment instances attached with user.
+            This only includes CourseAccessRole & CourseEnrollment instances related to
+            the provided CourseKey.
+
+    Returns:
+        (QuerySet): User queryset
+    """
+    queryset = User.objects.filter()
+
+    # if access_roles not given, assign all registered access roles
+    if access_roles is None:
+        access_roles = list(REGISTERED_ACCESS_ROLES.keys())
+
+    # conditions for filtering based on CourseAccessRole
+    access_role_qs = Q(
+        courseaccessrole__course_id=course_key,
+        courseaccessrole__role__in=access_roles
+    )
+
+    # conditions for filtering based on CourseEnrollment
+    students_qs = Q(
+        courseenrollment__course_id=course_key,
+        courseenrollment__is_active=True
+    )
+
+    if include_students:
+        queryset = queryset.filter(access_role_qs | students_qs)
+    else:
+        queryset = queryset.filter(access_role_qs)
+
+    if prefetch_user_course_roles:
+
+        # prefetch CourseAccessRole items related to the course and given roles
+        queryset = queryset.prefetch_related(Prefetch(
+            'courseaccessrole_set',
+            CourseAccessRole.objects.filter(
+                course_id=course_key, role__in=access_roles)
+        ))
+
+        # prefetch CourseEnrollment items related to the course
+        queryset = queryset.prefetch_related(Prefetch(
+            'courseenrollment_set',
+            CourseEnrollment.objects.filter(course_id=course_key)
+        ))
```

Comment:
@shimulch I don't think the behavior here is the expected one. We always want the `course_id` filter to be applied here. _I suggested the prefetching is just to avoid making repetitive queries when retrieving data from the related models,_ but I see that there's some filtering happening. 

I think the best approach here is to:
1. Remove the `prefetch_user_course_roles` conditional, leaving prefetching and filtering to always happen.
2. Change the function docstring to be explicitly clear that only course access roles and enrollments **related to the queried course will be retrieved**.
3. Update tests to take this into account, and check that accessing the related objects doesn't trigger extra queries when using this API.